### ISSUE TITLE
`Prio3`: What is the right abstraction boundary for measurements/results?

### ISSUE THREAD

**Issue by cjpatton** [divviup/libprio-rs#235]
As of #228, this our API for `Prio3`:
```rust
  impl<T, A, P, const L: usize> Vdaf for Prio3<T, A, P, L>
  where
      T: Type,
      A: Clone + Debug + Sync + Send,
      P: Prg<L>,
  {
      type Measurement = T::Measurement;
      type AggregateResult = A;
      type AggregationParam = ();
      type InputShare = Prio3InputShare<T::Field, L>;
      type OutputShare = OutputShare<T::Field>;
      type AggregateShare = AggregateShare<T::Field>;

      fn num_aggregators(&self) -> usize {
          self.num_aggregators as usize
      }
  }
```

There are two features of this API I want to discuss:
1. `Prio3::Measurement` is determined by the underlying `T::Measurement`. Typically this is either the integer representation of elements of the finite field (`T::Field`) or a vector of such integers. However, this design also envisions more complex measurement types, such as some fixed point numbers.
2. `Prio3::AggregateResult` is generic. Similar to measurement, typically the concrete type is either the integer representation of elements of the finite field or a vector of such integers. Again, this design envisions more complex aggregate results.

These design choices add quite a bit of API complexity to `Prio3`. Luckily this complexity doesn't bleed into concrete instantiations of `Prio3`, i.e., `Prio3Aes128Count`, `Prio3Aes128CountVec`, and so on. Still, I think it's worth asking ourselves if the design decisions of (1.) and (2.) drew the right abstraction boundary.

Suppose instead we defined both `Prio3::Mesurement` and `Prio3::AggregateResult` to be `Vec<T::Field::Integer>`. I claim that this is sufficiently general. However the downside is that this shifts to the user the burden of encoding measurements as `Vec<T::Field::Integer>` and decoding aggregate results from `Vec<T::Field::Integer>`. On the other hand I can see at least two upsides:
* Less complexity for `Prio3` API
* Supports code reuse. For example, suppose we wanted a variant of `Prio3Aes128Sum` that is defined for fixed-point numbers rather than integers. In theory the same code could be used for both, without having to define separate types for each.

**Comment by divergentdave** [comment]
I'm partial to the generality of being able to define these types. In our Typescript implementation, we made the measurement type of Prio3Aes128Count a boolean. One other note: if we have a sum of squares/variance variant of Prio3, that would probably need to be aware of where the fixed point radix is, in order to allow reuse. (Otherwise, you would need to trade off between very large field sizes and the sum of squares result having too much precision/not enough range)

**Comment by cjpatton** [comment]
Hey @divergentdave, I think I found a way to make us both happy: Instead of making agg result type generic, have it be defined by the underlying `Type`. See the linked PR.

**Comment by cjpatton** [comment]
Closed by #239.
